### PR TITLE
Nebharg/add api use region

### PR DESCRIPTION
**Pull request by neha-bhargava**
#2259

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/Region/RegionSource.cs [hunk 1]
```diff
@@ -9,6 +9,7 @@ internal enum RegionSource
         None = 0,
         EnvVariable = 1,
         Imds = 2,
-        Cache = 3
+        Cache = 3,
+        Userprovided = 4
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client/Region/RegionSource.cs:13]: `UserProvided`

File: src/client/Microsoft.Identity.Client/TelemetryCore/Internal/Events/ApiEvent.cs [hunk 2]
```diff
@@ -169,6 +170,12 @@ public int RegionSource
             set => this[RegionSourceKey] = (value).ToString(CultureInfo.InvariantCulture);
         }
 
+        public string ValidateUseRegion
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client/TelemetryCore/Internal/Events/ApiEvent.cs:173]: Hmm maybe rename to something like `IsValidUserProvidedRegion` and maybe make it a nullable bool?

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RegionalAuthIntegrationTests.cs [hunk 3]
```diff
@@ -205,10 +278,11 @@ public async Task AcquireTokenToGlobalEndpointThenRegionalEndpoint_UseTokenFromC
 
         private async Task<AuthenticationResult> CreateAuthenticationResultAsync(
```
Comment (pmaytak) [inline tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RegionalAuthIntegrationTests.cs:279]: Not part of this PR, but I would maybe rename this to `GetAuthenticationResultAsync`, so that it's clearer that we're calling a method on CCA and not creating a mock auth result object?

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RegionalAuthIntegrationTests.cs [hunk 4]
```diff
@@ -103,6 +103,73 @@ public async Task AcquireTokenToRegionalEndpointAsync()
                 AuthenticationResult result = await CreateAuthenticationResultAsync().ConfigureAwait(false); // regional endpoint
                 AssertTokenSource_IsIdP(result);
                 AssertValidHost(true, factory);
+                AssertTelemetry(factory, "2|1004,0|centralus,1,0,");
+            }
+            finally
+            {
+                Environment.SetEnvironmentVariable(TestConstants.RegionName, null);
```
Comment (pmaytak) [inline tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RegionalAuthIntegrationTests.cs:110]: Would it be appropriate to put this line into a separate test cleanup method, so it doesn't have to be repeated?

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs [hunk 5]
```diff
@@ -79,10 +79,13 @@ public AcquireTokenForClientParameterBuilder WithSendX5C(bool withSendX5C)
         /// By default the value for this variable is false.
         /// See https://aka.ms/msal-net-region-discovery for more details.
         /// </summary>
-        /// <param name="autoDetectRegion"><c>true</c> if the token request should be sent to regional ESTS. Otherwise <c>false</c>.
-        /// The default is <c>false</c></param>
+        /// <param name="autoDetectRegion"><c>true</c> if the token request should be sent to regional ESTS. The default is <c>false</c>.
+        /// </param>
+        /// <param name="useRegion"> optional parameter to provide region to MSAL. This parameter will be used along with auto detection of region.
+        /// If the region is auto detected, the provided region will be compared with the detected region and used in telemetry to do analysis on correctness of the region provided.
+        /// If auto region detection fails, the provided region will be used for instance metadata.</param>
```
[the remark this answers is not in the document]
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs:86]: The auto detect region is set to true if the user wants to opt for regional endpoint. If set to false or not set, the requests will go to the global endpoint as usual.
UseRegion is an optional api where a user can provide the region in case they have it set in their config or they know the region. In case we are able to detect the region we will use the auto detected region and compare it with the region provided by the user to analyse the correctness of the region provided. In case we fail to detect the region and we have a value in useRegion, we will go ahead and use this region to hit the regional endpoint. And if in case user provided some garbage value in that place and we form a regional endpoint using it, ests is working on a way to handle those cases.

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs [hunk 6]
```diff
@@ -79,10 +79,13 @@ public AcquireTokenForClientParameterBuilder WithSendX5C(bool withSendX5C)
...
         /// <returns>The builder to chain the .With methods</returns>
-        public AcquireTokenForClientParameterBuilder WithAzureRegion(bool autoDetectRegion)
+        public AcquireTokenForClientParameterBuilder WithAzureRegion(bool autoDetectRegion, string useRegion = "")
```
[the remark this answers is not in the document]
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs:88]: Makes sense, regionToUse sounds good. I will update it.

[the remark this answers is not in the document]
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs]: Updated

File: src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenForClientParameters.cs [hunk 7]
```diff
@@ -20,13 +20,18 @@ internal class AcquireTokenForClientParameters : IAcquireTokenParameters
         /// </summary>
         public bool AutoDetectRegion { get; set; }
 
+        /// <summary>
+        /// </summary>
+        public string UseRegion { get; set; }
```
[the remark this answers is not in the document]
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenForClientParameters.cs]: Updated

File: src/client/Microsoft.Identity.Client/Region/RegionSource.cs (hunk 1 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/Region/RegionSource.cs]: Updated

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RegionalAuthIntegrationTests.cs (hunk 3 above)
Comment (neha-bhargava) [reply tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RegionalAuthIntegrationTests.cs]: Updated

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RegionalAuthIntegrationTests.cs (hunk 4 above)
Comment (neha-bhargava) [reply tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RegionalAuthIntegrationTests.cs]: Updated

File: src/client/Microsoft.Identity.Client/TelemetryCore/Internal/Events/ApiEvent.cs (hunk 2 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/TelemetryCore/Internal/Events/ApiEvent.cs]: Good point. Updated.

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs]: Why not just use the region, if it's provided, and skip the auto detect region call? Then, if the the call fails, I guess try to autoDetect the region and try the request again? Is this specific guidance from eSTS on how to handle this?

---
In reply to: [549887878](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2315#discussion_r549887878) [](ancestors = 549887878)

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs]: i'm a little confused...so sorry for the weird questions...so if I do `.WithAzureRegion()` does that mean I am expecting to use the regional endpoint? If there a reason to even have the boolean? If a customer includes `.WithAzureRegion()` can we assume they want the regional endpoint? so if a customer uses this, it's always going to regional, and then they can specify the actual region they want, if they know it. It just seems counterintuitive to say I want to use the regional endpoint, and then also have to set a boolean to true. interested in discussing offline if you want.

---
In reply to: [549874926](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2315#discussion_r549874926) [](ancestors = 549874926)

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs [hunk 8]
```diff
@@ -93,6 +96,7 @@ public AcquireTokenForClientParameterBuilder WithAzureRegion(bool autoDetectRegi
 
             CommonParameters.AddApiTelemetryFeature(ApiTelemetryFeature.WithAzureRegion, autoDetectRegion);
             Parameters.AutoDetectRegion = autoDetectRegion;
+            Parameters.RegionToUse = regionToUse;
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs:99]: if it's an empty string do we still want to set it, as that's the default?

File: src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenForClientParameters.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenForClientParameters.cs]: or maybe `RegionSpecifiedInConfig` 

---
In reply to: [550358060](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2315#discussion_r550358060) [](ancestors = 550358060)

File: src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs [hunk 9]
```diff
@@ -204,11 +206,31 @@ private void CacheInstanceDiscoveryMetadata(InstanceDiscoveryMetadataEntry metad
...
+
             if (s_region.IsNullOrEmpty())
             {
-                s_region = await GetRegionAsync(requestContext).ConfigureAwait(false);
+                try
+                {
+                    s_region = await GetRegionAsync(requestContext).ConfigureAwait(false);
+
+                    if (!regionToUse.IsNullOrEmpty())
+                    {
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs:218]: i would think the customer would know the region they want to target, and we should just use the provided value. what does ests suggest we do or have they provided input on that?

Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs:218]: why do we care if the region they provided is valid or not? like you mentioned earlier, they could provide a garbage region and then we are logging all these errors, but we have no telemetry dashboard set up to get the data on this. Why does ests care if the region provided by the customer matches the region we discover? seems like a test they should have on their side, but not necessarily something we need in MSAL.

---
In reply to: [551518052](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2315#discussion_r551518052) [](ancestors = 551518052)

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs]: This is required to analyze the validity of the value that user provides. We know that there are azure services where we cannot detect the region and there can be more platforms where auto detection fails. For this reason it was decided to allow the user to provide the region. But this is something we cannot completely rely on as it could be hardcoded or the value of region they provide might not be the actual azure region. So, we will analyze the validity of the region provided by the user and in the initial phases of adoption we can let them know that there is a mismatch in the region they provide and region that was detected. Also, in case we fail to compare we are logging the region and region source which can help to ensure the region provided is correct and the calls are going to regional endpoint. If user opts to use regional endpoint goal is to hit the regional endpoint and use global only as a last option.

---
In reply to: [551511449](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2315#discussion_r551511449) [](ancestors = 551511449,549887878)

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs]: Initially, there was no plan to ask user to provide the region. But since now adopting regional is a resilience feature, we are adding things like allow the user to pass region as currently we cannot detect region on all platforms. So, the bool `autoDetectRegion` is what currently exists and people are testing it on their test environments. `regionToUse` is an optional field that someone may choose to provide and we will use it in case we are unable to auto detect the region to give more confidence to people adopting it.

---
In reply to: [551514796](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2315#discussion_r551514796) [](ancestors = 551514796,549874926)

File: src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs (hunk 9 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs:218]: It is highly likely that people use a value for region that is different from the actual azure region and with allowing user to provide the region we are also open to the possibility of someone hardcoding this value and hitting the wrong regional endpoint. Happy to discuss this offline.

---
In reply to: [551519474](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2315#discussion_r551519474) [](ancestors = 551519474,551518052)

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs]: right, that makes sense. i guess what i'm suggesting is simplifying the API to `public AcquireTokenForClientParameterBuilder WithAzureRegion(string regionToUse = "")` because if they use this builder, they are automatically opting in to regional detection, but if they provide a value, we 1) try that region 2) if that fails try to detect the region 3) use global endpoint

---
In reply to: [551575530](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2315#discussion_r551575530) [](ancestors = 551575530,551514796,549874926)

File: src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs (hunk 9 above)
Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs:218]: highly likely or highly unlikely? if highly likely, then sounds like we should do something else to detect this sooner. also, customers won't have the telemetry logs to see the azure regions, is this also getting logged in the regular logs as well? I guess, how will they know which region they are hitting? +1 to discussing. maybe at stand-up tomorrow.

---
In reply to: [551576965](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2315#discussion_r551576965) [](ancestors = 551576965,551519474,551518052)

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jmprieur) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs]: Thanks for the explanation, @neha-bhargava, I had a similar question as @jennyf19 

In that case, what about having the following public API: `WithAzureRegion(string regionToUse="autoDetect")`:
- If the developer provides the region, then for sure the boolean would be true ? 
- and if they don't provide it (then the region would become) "autoDetect", which you'd then internally set boolean `autoDetectRegion` to true (but that would not be needed to be surfaced to the developers)

Note: you don't need to change your current method. Just make it internal, and wrap it into the API I propose?

I was also wondering if there is a value of passing `WithAzureRegion(false)`? What is the use case? would it be that the developer used the regional endpoint and then want to go to Global? but given this is on the AcquireTokenForClientParameterBuilder, the developer could just omit the `.WithAzureRegion()` to do that?

File: src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs (hunk 9 above)
Comment (jmprieur) [reply src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs:218]: Do we want to have an API on the IConfidentialClientApplication which would be string GetAzureRegion() ?

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (pmaytak) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs]: Side note: We could just check for a default empty string, `WithAzureRegion(string regionToUse="")`. Would perhaps changing the parameter name to `fallbackRegion` be clearer?

[the remark this answers is not in the document]
Comment (henrik-me) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs]: @jmprieur : the reason to have the boolean is to avoid developers having to write if statements e.g.
if (boolUseRegion)
  builder.withRegion(...)

instead they can write withyRegion(boolUseRegion, "valueFromMyConfig")

#CleanerCode

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (henrik-me) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs]: I think you should revert back to what you had before for the reason mentioned above. the developer experience is better that way as it avoids the need for if statements

---
In reply to: [550357996](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2315#discussion_r550357996) [](ancestors = 550357996)

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs [hunk 10]
```diff
@@ -93,6 +96,7 @@ public AcquireTokenForClientParameterBuilder WithAzureRegion(bool autoDetectRegi
 
             CommonParameters.AddApiTelemetryFeature(ApiTelemetryFeature.WithAzureRegion, autoDetectRegion);
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs:97]: >CommonParameters.AddApiTelemetryFeature(ApiTelemetryFeature.WithAzureRegion, autoDetectRegion); [](start = 12, length = 95)

Wondering if you should set the regionToUse value in telemetry here as well

File: src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenForClientParameters.cs [hunk 11]
```diff
@@ -20,13 +20,18 @@ internal class AcquireTokenForClientParameters : IAcquireTokenParameters
         /// </summary>
         public bool AutoDetectRegion { get; set; }
 
+        /// <summary>
+        /// </summary>
+        public string RegionToUse { get; set; }
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenForClientParameters.cs:25]: Even though this is an internal class, I believe the comments made on what the data is warrents that the summary field is specified.

File: src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs [hunk 12]
```diff
@@ -204,11 +206,31 @@ private void CacheInstanceDiscoveryMetadata(InstanceDiscoveryMetadataEntry metad
...
         private async Task<Uri> BuildAuthorityWithRegionAsync(Uri canonicalAuthority, RequestContext requestContext)
         {
+            string regionToUse = requestContext.ServiceBundle.Config.AuthorityInfo.RegionToUse;
+
             if (s_region.IsNullOrEmpty())
             {
-                s_region = await GetRegionAsync(requestContext).ConfigureAwait(false);
+                try
+                {
+                    s_region = await GetRegionAsync(requestContext).ConfigureAwait(false);
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs:215]: >s_region [](start = 20, length = 8)

what happens if this is null?

File: src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs [hunk 13]
```diff
@@ -204,11 +206,31 @@ private void CacheInstanceDiscoveryMetadata(InstanceDiscoveryMetadataEntry metad
...
+
+                    if (!regionToUse.IsNullOrEmpty())
+                    {
+                        requestContext.ApiEvent.IsValidUserProvidedRegion = s_region.Equals(regionToUse);
+                    }
+                }
+                catch (MsalServiceException e)
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs:222]: >e [](start = 44, length = 1)

you are not using this, remove and simply throw instead

File: src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs [hunk 14]
```diff
@@ -204,11 +206,31 @@ private void CacheInstanceDiscoveryMetadata(InstanceDiscoveryMetadataEntry metad
...
+                {
+                    if (regionToUse.IsNullOrEmpty())
+                    {
+                        throw e;
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs:226]: >e [](start = 30, length = 1)

remove

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs]: @henrik-me how would that lead to not using if statements? if they don't want to use the region, then they just won't use .withRegion(). in the AT call, right?

File: src/client/Microsoft.Identity.Client/TelemetryCore/Http/HttpTelemetryManager.cs [hunk 15]
```diff
@@ -120,16 +120,18 @@ public string GetCurrentRequestHeader(ApiEvent eventInProgress)
...
+            var platformConfig = new StringBuilder();
+            platformConfig.Append(regionDiscovered + ",");
+            platformConfig.Append(regionSource + ",");
+            platformConfig.Append(ConvertFromStringToBitwise(isTokenCacheSerialized) + ",");
+            platformConfig.Append(string.IsNullOrEmpty(isValidUserProvidedRegion) ? isValidUserProvidedRegion : ConvertFromStringToBitwise(isValidUserProvidedRegion)); //The value for this will be 1 if the region provided is valid, 0 if invalid and "" in case it could not be validated.
 
             return $"{TelemetryConstants.HttpTelemetrySchemaVersion2}" +
                 $"|{apiId},{ConvertFromStringToBitwise(forceRefresh)}" +
-                $"|{regionalFields},{ConvertFromStringToBitwise(isTokenCacheSerialized)}";
+                $"|{platformConfig}";
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/TelemetryCore/Http/HttpTelemetryManager.cs:134]: > $"|{platformConfig}"; [](start = 15, length = 22)

can you pls. show a few examples of what this looks like from your testing? with region, without region specified, where region can be discovered, where it can't be discovered, etc?

File: src/client/Microsoft.Identity.Client/TelemetryCore/Internal/Events/ApiEvent.cs [hunk 16]
```diff
@@ -169,6 +170,24 @@ public int RegionSource
             set => this[RegionSourceKey] = (value).ToString(CultureInfo.InvariantCulture);
         }
 
+        public bool? IsValidUserProvidedRegion
+        {
+#pragma warning disable CA1305 // .net standard does not have an overload for ToString() with Culture
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/TelemetryCore/Internal/Events/ApiEvent.cs:175]: >.net standard does not have an overload for ToString() with Culture [](start = 34, length = 67)

aside: [https://docs.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings?redirectedfrom=MSDN#stringtoupper-and-stringtolower](https://docs.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings?redirectedfrom=MSDN#stringtoupper-and-stringtolower)

File: tests/Microsoft.Identity.Test.Unit/TelemetryTests/RegionalTelemetryTests.cs [hunk 17]
```diff
@@ -138,9 +137,86 @@ public async Task TelemetrySerializedTokenCacheTestAsync()
...
+                    ApiIds.AcquireTokenForClient,
+                    forceRefresh: false,
+                    "1",
+                    isCacheSerialized: false,
+                    isvalidUserProvidedRegion: "1");
+                AssertPreviousTelemetry(result.HttpRequest, expectedSilentCount: 0);
+            }
+            finally
+            {
+                Environment.SetEnvironmentVariable(TestConstants.RegionName, "");
```
Comment (henrik-me) [inline tests/Microsoft.Identity.Test.Unit/TelemetryTests/RegionalTelemetryTests.cs:183]: >Environment.SetEnvironmentVariable(TestConstants.RegionName, ""); [](start = 16, length = 65)

move to test cleanup? similar to what you did in the other test class?

**Review by henrik-me — APPROVED** [review]
:shipit:

File: src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs (hunk 12 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs:215]: Yes, in case the region is not detected an exception is thrown

File: src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs (hunk 13 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs:222]: This is to catch the exception if the region was not detected. If in this case we don't even have a user provided region we throw the error and let the fallback to global logic handle it later. In case we have a user provided region, we don't throw an error but use that region.

File: src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs (hunk 14 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs:226]: This ensures that we only throw an exception when we fail to auto detect and we don't have a user provided region.

File: src/client/Microsoft.Identity.Client/TelemetryCore/Internal/Events/ApiEvent.cs (hunk 16 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/TelemetryCore/Internal/Events/ApiEvent.cs:182]: If I use toUpper, it messes up with code [here](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/c98af5f29b76077a9bf596b3e0c0dc5d78244755/src/client/Microsoft.Identity.Client/TelemetryCore/Http/HttpTelemetryManager.cs#L132) when it converts the bool to bitwise. So leaving it as is.

File: tests/Microsoft.Identity.Test.Unit/TelemetryTests/RegionalTelemetryTests.cs (hunk 17 above)
Comment (neha-bhargava) [reply tests/Microsoft.Identity.Test.Unit/TelemetryTests/RegionalTelemetryTests.cs]: Done

File: src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenForClientParameters.cs (hunk 11 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenForClientParameters.cs:27]: Added

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs (hunk 10 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs:97]: Added this to telemetry

**Comment by neha-bhargava** [comment]
Kusto query and results for different scenarios with user provided region.

Format in MsalCur: 2|api_id,force_refresh|region,region_source,is_token_cache_serialized,user_provided_region,validate_use_region

AllPerRequestTable
| where CorrelationId == "994f2694-f88f-40da-baa4-5cf887106fa7"
| where env_time between(datetime(2021-01-07 18:00:00) .. datetime(2021-01-07 20:55:35))
| project ApplicationId, env_time, CorrelationId, ClientInfo, MsalCur, MsalLast

Scenario 1: No user provided region, auto detection successful
Correlation id: 994f2694-f88f-40da-baa4-5cf887106fa7
image

Scenario 2: Valid User provided region, auto detection successful
Correlation id: 95eb45ca-0b95-42e2-b161-fda7371e4909
image

Scenario 3: Invalid user provided region, auto detection successful
Correlation id: 8339cf22-e597-4c53-b8e4-011d4b8d373c
image

Scenario 4: User provided region, auto detection fails
Correlation id: 8ea300e6-12b0-40fd-81cc-f7f9a6f95dda
image

Global
Correlation id: a4f35dab-a957-4505-b29e-f89ee616ff5b
image

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs (hunk 8 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForClientParameterBuilder.cs:99]: changed the default value to null.

File: src/client/Microsoft.Identity.Client/TelemetryCore/Http/HttpTelemetryManager.cs (hunk 15 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/TelemetryCore/Http/HttpTelemetryManager.cs:136]: Added a comment with scenarios and kusto query results.
